core/src/main/java/org/elasticsearch/search/fetch/subphase/DocValueFieldsFetchSubPhase.java

```diff
@@ -48,26 +53,42 @@ public void hitExecute(SearchContext context, HitContext hitContext) throws IOEx
                 context.docValueFieldsContext().fields().add(name);
             }
         }
+
         if (context.docValueFieldsContext() == null) {
             return;
         }
+
+        hits = hits.clone(); // don't modify the incoming hits
+        Arrays.sort(hits, (a, b) -> Integer.compare(a.docId(), b.docId()));
```

Comment:
matter of taste but I tend to like method refs better, ie. `Arrays.sort(hits, Comparators.comparing(SearchHit::docId))`